### PR TITLE
[NFC][SYCLomatic][DOC]Add FAQ for synchronization in loop/condition statement

### PR DESCRIPTION
**Pull request by wenhuiNi**
Signed-off-by: Ni, Wenhui <wenhui.ni@intel.com>

### CODE REVIEW

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 1]
```diff
@@ -162,6 +163,152 @@ expected results. Instead, exclude CUB library implementation source code from
 your migration by adding ``--in-root-exclude=<path to CUB library source code>``
 to your migration command.
 
+How do I fix the migrated code in case barrier is used conditional statement?
```
Comment (AndyCHHuang) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: "in" conditional statement

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 2]
```diff
@@ -162,6 +163,152 @@ expected results. Instead, exclude CUB library implementation source code from
...
+to ensure that all work_items within a sub_group must be encountered synchronization 
+points in converged control flow.
+
+Here are some examples.
+Barrier is used in conditional statements. The following original code:
+
+.. code-block:: cpp
+   :linenos:
+
+   // original code
```
Comment (zhiweij1) [inline docs/dev_guide/frequently-asked-questions-faq.rst:184]: original code ==> migrated code?

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 3]
```diff
@@ -162,6 +163,152 @@ expected results. Instead, exclude CUB library implementation source code from
...
+         if (sdata[tid] < sdata[tid + 1]) {
+            ind[tid] = ind[tid + 1];
+            sdata[tid] = sdata[tid + 1];
+         }
+      }
+      item_ct1.barrier(sycl::access::fence_space::local_space);
+      }
+   }
+
+Barrier is used in loop statements. The following original code:
```
Comment (zhiweij1) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: original code ==> migrated code?

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 4]
```diff
@@ -162,6 +163,152 @@ expected results. Instead, exclude CUB library implementation source code from
...
+How do I fix the migrated code in case barrier is used conditional statement?
+*****************************************************************************
+
+If the migration code you are running uses synchronization within conditional/loop
+statements, you may encounter issues like hang or runtime_error. It is essential
+to ensure that all work_items within a sub_group must be encountered synchronization 
+points in converged control flow.
+
+Here are some examples.
+Barrier is used in conditional statements. The following original code:
```
Comment (AndyCHHuang) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: In the first example, nd_item.barrier() is used inside a if block and the evaluation results of the conditional statement differs in each thread. The following is the original code:

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 5]
```diff
@@ -162,6 +163,152 @@ expected results. Instead, exclude CUB library implementation source code from
 your migration by adding ``--in-root-exclude=<path to CUB library source code>``
 to your migration command.
 
+How do I fix the migrated code in case barrier is used conditional statement?
+*****************************************************************************
+
+If the migration code you are running uses synchronization within conditional/loop
```
Comment (AndyCHHuang) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: If the original code you are migrating calls sychronization APIs within control flow statements like conditional/loop, you may encounter runtime issues like a process hang with the migrated code. It is essential to ensure that each sychronization point is either skipped by all the work_items or visited by all the work_items in the same sub_group. If a sychronization point is visited by some work_items and skiped(due to different evaluation result of the predicate statement) by other work_items, the queue will stuck at current task and cause the process hang.

**Review by zhiweij1 — APPROVED** [review]

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 6]
```diff
@@ -162,6 +163,152 @@ expected results. Instead, exclude CUB library implementation source code from
...
+         item_ct1.barrier(sycl::access::fence_space::local_space);
+         if (sdata[tid] < sdata[tid + 1]) {
+            ind[tid] = ind[tid + 1];
+            sdata[tid] = sdata[tid + 1];
+         }
+         item_ct1.barrier(sycl::access::fence_space::local_space);
+      }
+   }
+
+is adjusted to move the synchronization statement outside of the
```
Comment (AndyCHHuang) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: The following manually-adjusted code shows how to fix the hang issue by moving the synchronization statement out of the if block.

File: docs/dev_guide/frequently-asked-questions-faq.rst (hunk 3 above)
Comment (AndyCHHuang) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: The second example demonstrates how to fix the hang issue when a synchronization statement is used in a for loop in the original code:

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 7]
```diff
@@ -162,6 +163,152 @@ expected results. Instead, exclude CUB library implementation source code from
...
+         item_ct1.get_local_id(2);
+      for (; curr_particle < np; curr_particle +=
+         item_ct1.get_group_range(2) * item_ct1.get_local_range(2)) {
+         ...
+         item_ct1.barrier();
+         ...
+      }
+   }
+
+is adjusted to fix the conditional in loop statement, let all thread
```
Comment (AndyCHHuang) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: The following is the manually-adjusted code of moving the synchronization statement out of the for loop.

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by ziranzha — APPROVED** [review]

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 8]
```diff
@@ -162,6 +163,160 @@ expected results. Instead, exclude CUB library implementation source code from
 your migration by adding ``--in-root-exclude=<path to CUB library source code>``
 to your migration command.
 
+How do I fix the sycl code in case synchronized block is used in conditional statement?
```
Comment (zhimingwang36) [inline docs/dev_guide/frequently-asked-questions-faq.rst]:
```suggestion
How do I fix the hang issue of sycl code cased by work group level synchronization like group barrier used in conditional statement?

```

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 9]
```diff
@@ -162,6 +163,98 @@ expected results. Instead, exclude CUB library implementation source code from
...
+      unsigned int tid = item_ct1.get_local_id(2);
+      if (tid < 32) {
+         ...
+         item_ct1.barrier(sycl::access::fence_space::local_space);
+         ...
+      }
+   }
+
+The following manually-adjusted code shows how to fix the hang issue
+or crash by moving the synchronization statement out of the if block.
```
Comment (zhimingwang36) [inline docs/dev_guide/frequently-asked-questions-faq.rst]:
```suggestion
by moving the synchronization statement out of the if block.

```

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 10]
```diff
@@ -162,6 +163,98 @@ expected results. Instead, exclude CUB library implementation source code from
...
+   void kernel(const sycl::nd_item<3> &item_ct1) {
+      unsigned int tid = item_ct1.get_local_id(2);
+      if (tid < 32) {
+         ...
+         item_ct1.barrier(sycl::access::fence_space::local_space);
+         ...
+      }
+   }
+
+The following manually-adjusted code shows how to fix the hang issue
```
Comment (zhimingwang36) [inline docs/dev_guide/frequently-asked-questions-faq.rst]:
```suggestion
The following code shows how to fix the hang issue

```

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 11]
```diff
@@ -162,6 +163,98 @@ expected results. Instead, exclude CUB library implementation source code from
...
+
+   // fixed SYCL code
+
+   void compute(int id_space, const sycl::nd_item<3> &item_ct1) {
+      unsigned int id =
+         item_ct1.get_group(2) * item_ct1.get_local_range(2) +
+         item_ct1.get_local_id(2);
+      unsigned int num_workitem = 
+               item_ct1.get_group_range(2) * item_ct1.get_local_range(2);
+      // The condition is rounded up to an integer multiple of the num of work items
```
Comment (zhimingwang36) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: the condition is updated to make sure all work items can enter the loop body in each iteration

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 12]
```diff
@@ -162,6 +163,98 @@ expected results. Instead, exclude CUB library implementation source code from
...
+         item_ct1.get_local_id(2);
+      for (; id < id_space; id +=
+         item_ct1.get_group_range(2) * item_ct1.get_local_range(2)) {
+         ...
+         item_ct1.barrier();
+         ...
+      }
+   }
+
+The following is the manually adjusted code which ensures all work
```
Comment (zhimingwang36) [inline docs/dev_guide/frequently-asked-questions-faq.rst]:
```suggestion
The following code shows how to fix the hang issue by making sure all work

```

**Review by zhimingwang36 — APPROVED** [review]
LGTM,  please ask DOC-team to review.

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 13]
```diff
@@ -11,6 +11,7 @@ Frequently Asked Questions
 * `How do I use the migrated module file in the new project?`_
 * `Is the memory space allocated by sycl::malloc_device, sycl::malloc_host, and dpct::dpct_malloc initialized?`_
 * `How do I migrate CUDA\* source code that contains CUB library implementation source code?`_
+* `How do I fix the hang issue of sycl code caused by work group level synchronization like group barrier used in conditional statement?`_
```
Comment (bdmoore1) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: How do I fix the issue of SYCL* code hanging due to work group level synchronization, such as a group barrier used in a conditional statement?

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 14]
```diff
@@ -162,6 +163,95 @@ expected results. Instead, exclude CUB library implementation source code from
 your migration by adding ``--in-root-exclude=<path to CUB library source code>``
 to your migration command.
 
+How do I fix the hang issue of sycl code caused by work group level synchronization like group barrier used in conditional statement?
```
Comment (bdmoore1) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: How do I fix the issue of SYCL* code hanging due to work group level synchronization, such as a group barrier used in a conditional statement?

File: docs/dev_guide/frequently-asked-questions-faq.rst [hunk 15]
```diff
@@ -162,6 +163,95 @@ expected results. Instead, exclude CUB library implementation source code from
 your migration by adding ``--in-root-exclude=<path to CUB library source code>``
 to your migration command.
 
+How do I fix the hang issue of sycl code caused by work group level synchronization like group barrier used in conditional statement?
+*************************************************************************************************************************************
+
+In SYCL code, if there is calling of synchronization API in control flow statements
```
Comment (bdmoore1) [inline docs/dev_guide/frequently-asked-questions-faq.rst]: If synchronization API in control flow statements like a conditional statement and loop statement are called in SYCL code, you may encounter a runtime hang issue.
The basic idea to fix the hang issue is to ensure that each synchronization API is
either reached by all work items of a workgroup, or skipped by all the work items of
a workgroup.

Here are two examples of how to fix: 

In the first example, the synchronization API group barrier(nd_item.barrier()) is called
inside an if block. The evaluation results of the conditional statement differs in
each work item so not all work items can reach the synchronization API.

**Review by bdmoore1 — COMMENTED** [review]
Added some changes to make the instructions more clear. Please review my suggestions to ensure they are still technically accurate.

File: docs/dev_guide/frequently-asked-questions-faq.rst (hunk 15 above)
Comment (wenhuiNi) [reply docs/dev_guide/frequently-asked-questions-faq.rst]: has fixed

File: docs/dev_guide/frequently-asked-questions-faq.rst (hunk 13 above)
Comment (wenhuiNi) [reply docs/dev_guide/frequently-asked-questions-faq.rst]: Add a backslash before*, It seem that "*" would be used as a control character.
Please double conform.

File: docs/dev_guide/frequently-asked-questions-faq.rst (hunk 14 above)
Comment (wenhuiNi) [reply docs/dev_guide/frequently-asked-questions-faq.rst]: same of the above

File: docs/dev_guide/frequently-asked-questions-faq.rst (hunk 15 above)
Comment (wenhuiNi) [reply docs/dev_guide/frequently-asked-questions-faq.rst]: I made a change
"The evaluation results of the conditional statement differs in" ==> "The evaluation results of the conditional statement differ in"
please double check. Thanks

File: docs/dev_guide/frequently-asked-questions-faq.rst (hunk 13 above)
Comment (bdmoore1) [reply docs/dev_guide/frequently-asked-questions-faq.rst]: Looks good to me.

**Pull request merged**